src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                                              "dependencies due to "
+                                              "taskloop dependencies "
+                                              "across barrierless OMP "
+                                              "serial regions.")
+
+    @staticmethod
+    def get_forward_dependence(taskloop, root):
+        '''
+        Returns the next forward dependence for a taskloop using the new
+        dependency analysis functions.
+        Forward dependencies can be of the following types:
+        Loop
+        OMPDoDirective
+        OMPTaskloopDirective
+        OMPTaskwaitDirective (If in same OMPSingle and that single has
+        nowait=False)
+        OMPSingleDirective (If ancestor OMPSingle has nowait=False)
+        OMPParallelDirective
+
+        The forward dependency is never a child of taskloop, and must have
+        abs_position > taskloop.abs_position
+
+        :param taskloop: the taskloop node to search for the
+                         forward_dependence.
+        :type taskloop: :py:class:`psyclone.psyir.nodes.OMPTaskloopDirective
+        :param root: the tree in which to search for the forward_dependence.
+        :type root: :py:class:`psyclone.f2pygen.DoGen`
+
+        :returns: the forward_dependence of taskloop.
+        :rtype: :py:class:`psyclone.f2pygen.DoGen`
+        '''
+        # We only look for specific types
+        node_list = root.walk((Loop, OMPDoDirective, OMPTaskloopDirective,
+                               OMPTaskwaitDirective, OMPSingleDirective,
+                               OMPParallelDirective))
+        taskloop_vars = VariablesAccessInfo()
+        for child in taskloop.walk(nodes.Node):
+            if child is not taskloop and not isinstance(child,
+                                                        (Schedule, Loop)):
+                taskloop_vars.merge(VariablesAccessInfo(child))
+        taskloop_signatures = taskloop_vars.all_signatures
+        # Find our parent serial region if it has a barrier
+        parent_single = taskloop.ancestor(OMPSingleDirective)
+        if parent_single is not None and parent_single.nowait:
+            parent_single = None
+        # Find our parent parallel region
+        parent_parallel = taskloop.ancestor(OMPParallelDirective)
+        for node in node_list:
+            if node.abs_position <= taskloop.abs_position:
+                continue
+            # Ignore our children
+            anc = node.ancestor(OMPTaskloopDirective)
+            if anc is taskloop:
+                continue
+            node_vars = None
+            # Special case:
+            # 1. If we have a different parent parallel node, then our parent
+            # parallel node is our dependency
+            if node.ancestor(OMPParallelDirective) is not parent_parallel:
+                return parent_parallel
+
+            if (isinstance(node, OMPTaskwaitDirective) and
```

Comment:
Bearing in mind that going up and down the tree can get expensive, please could you cache some of these results (e.g. `taskloop.ancestor(OMPSingleDirective)`) so that we find it just once and then re-use it. It might also help make this `if` a bit shorter :-)

Reply:
I think I've cached the ones I can, unfortunately the code checks for different types of ancestor a lot :(

Reply:
I checked `ancestor` docstring and it says it only returns the first instance found. I didn't check the code explicitly so its possible thats wrong.

Reply:
No idea what I was thinking here :(  Please ignore ;)